Add Withdrawal Pattern Example

Denton-L opened the pull request:
I feel like this is very rough, currently. Please let me know about what improvements I can make with this!

resolve #705

Denton-L committed:
Add section about withdrawal pattern

Denton-L commented:
I have no idea why the build is failing right now.

Denton-L committed:
Link to withdraw pattern

VoR0220 commented:
glitch in the matrix I suppose

chriseth commented on docs/common-patterns.rst:
```diff
@@ -2,6 +2,82 @@
+*************************
+
+The recommended method of sending funds after an effect
+is with the withdrawal pattern. Although the most intuitive
```
`is using the`?

chriseth commented on docs/common-patterns.rst:
```diff
@@ -2,6 +2,82 @@
+
+The recommended method of sending funds after an effect
+is with the withdrawal pattern. Although the most intuitive
+aethod of sending Ether as a result of an effect is a
```
`method`

chriseth commented on docs/common-patterns.rst:
```diff
@@ -2,6 +2,82 @@
+        }
+    }
+
+This is as opposed to the more intuitive sending pattern.
```
We could add that it is fine in this extremely simple example, but gets harder and harder to handle correctly with more complex contracts.

chriseth commented on docs/common-patterns.rst:
```diff
@@ -2,6 +2,82 @@
+        mapping (address => uint) pendingReturns;
+
+        function sendEther(uint amount) {
+            if (amount < etherStore[msg.sender]) {
```
`amount > etherStore[...`

chriseth commented on docs/common-patterns.rst:
```diff
@@ -2,6 +2,82 @@
+
+        mapping (address => uint) etherStore;
+
+        function sendEther(uint amount) {
```
This is basically also the withdraw pattern. I think we should create a more complicated example where we have an actual effect.

Denton-L committed:
Fix withdrawal pattern documentation

Denton-L commented:
@chriseth, do you have any suggestions on what kind of example might be a good one to use?

chriseth commented:
@Denton-L the auction example already has the withdrawal pattern, perhaps we could borrow a snippet from there?

Denton-L committed:
Remove trailing whitespace

Denton-L committed:
Change example to auction

Denton-L commented:
@chriseth, I have included a modified version of the auction example.

chriseth commented:
Sorry, but I'm not really keen on just repeating the auction example here. If the pattern cannot be shown properly in isolation, then we should rather link to the auction example.

Denton-L committed:
Use store example

Denton-L commented:
Sorry, I've been busy for the past little bit. I've updated it so that it uses a different example, the "store" example. This should be shorter and simpler than the auction example.

chriseth commented:
Thanks for the update. Sorry to say that, but I think that also this is not the best example. Examples where you really need the withdrawal patterns are those where the person triggering the action does not necessarily benefit from the transfer.

In the example here, you could just use `if(!x.send(y)) throw` - and in fact you should because otherwise this example will generate a compiler warning. Using `throw` does not harm here, because it does not stall a process. Where you cannot use throw and have to use withdraw is where the sending is part of another process that can then be stalled by the person receiving the funds.

Denton-L commented:
Would a [King of the Ether](https://www.kingoftheether.com/)-type example work, then?

It acts as almost an Auction contract without as much complexity.

chriseth commented:
Yes, that's a good idea!

Denton-L committed:
Change withdrawal example

The example is now a "King of the Ether"-esque contract. This is
actually relevant as they suffered an attack because of an almost
identical issue.

See the post-mortem here:
https://www.kingoftheether.com/postmortem.html

Denton-L committed:
Update contracts and descriptions

Denton-L committed:
Add minor corrections

chriseth commented on docs/common-patterns.rst:
```diff
@@ -2,6 +2,105 @@
+introduces a potential security risk. You may read
+more about this on the :ref:`security_considerations` page.
+
+This is an example of the withdrawal pattern in practice in
```
I think we should mention "king of the ether" as insipration here.

chriseth commented on docs/common-patterns.rst:
```diff
@@ -2,6 +2,105 @@
+        address public richest;
+        uint public mostSent;
+
+        mapping (address => uint) pending;
```
`pendingWithdrawal`?

chriseth commented on docs/common-patterns.rst:
```diff
@@ -2,6 +2,105 @@
+
+        function becomeRichest() returns (bool) {
+            if (msg.value > mostSent) {
+                pending[richest] = msg.value;
```
I think we need `pending[richest] += msg.value` here.

Denton-L committed:
Reference inspiration

Denton-L committed:
Fix code

Denton-L committed:
Remove throw from withdrawal pattern

Denton-L commented:
@chriseth, as in #913, I changed the example such that it does not throw anymore.

chriseth commented on docs/common-patterns.rst:
```diff
@@ -2,6 +2,110 @@
+
+        function becomeRichest() returns (bool) {
+            if (msg.value > mostSent) {
+                pendingWithdrawals[richest] += msg.value;
```
I think it should be `pendingWithdrawals[richest] += mostSent` here.

Denton-L replied:
I think the point of this is that your profit is the difference of whatever the person after sent in. That way, each person makes a little bit of money except for the very last person to send in their Ether.

chriseth replied:
Ah, it seems I completely misunderstood that example :-)

chriseth commented on docs/common-patterns.rst:
```diff
@@ -2,6 +2,110 @@
+                // Check if call succeeds to prevent an attacker
+                // from trapping the previous person's funds in
+                // this contract through a callstack attack
+                if (!richest.send(msg.value)) {
```
`richest.send(mostSent)`

Denton-L replied:
See other comment.
